### ISSUE TITLE
RR can get stuck in an infinite loop during reverse-step

### ISSUE THREAD

**Issue by gaasedelen** [mozilla/rr#2143]
Filing this so it doesn't get lost/forgotten. I haven't had a chance to investigate this.

## Overview

There appears to be a bug where rr can get stuck in an infinite loop while attempting to complete a `reverse-step` request. Based on the repro steps, the bug may stem from the use of hardware breakpoints while stepping around or over syscalls. 

Demo of the issue: https://asciinema.org/a/1pu9AI7IQHYUn4TcWIbpP36A4

## Stack Trace

I attached to the stuck proc and dumped its stack trace. Seems stuck in a loop.

```
(gdb) bt
#0  0x00007fa2e03c3dee in ptrace (request=32) at ../sysdeps/unix/sysv/linux/ptrace.c:45
#1  0x0000000000522728 in rr::Task::fallible_ptrace (this=0xee6340, request=32, addr=..., data=0x0) at ../src/Task.cc:1922
#2  rr::Task::ptrace_if_alive (this=0xee6340, request=32, addr=..., data=0x0) at ../src/Task.cc:2259
#3  0x000000000051edcc in rr::Task::resume_execution (this=0xee6340, how=rr::RESUME_SYSEMU_SINGLESTEP, wait_how=<optimized out>, tick_period=rr::RESUME_UNLIMITED_TICKS, sig=0) at ../src/Task.cc:1049
#4  0x000000000045e243 in rr::fast_forward_through_instruction (t=0xee6340, how=rr::RESUME_SYSEMU_SINGLESTEP, states=std::vector of length 1, capacity 1 = {...}) at ../src/fast_forward.cc:164
#5  0x00000000004e871b in rr::ReplaySession::cont_syscall_boundary (this=0xea9ae0, t=0xee6340, constraints=...) at ../src/ReplaySession.cc:357
#6  0x00000000004e8e20 in rr::ReplaySession::enter_syscall (this=0xea9ae0, t=0xee6340, constraints=...) at ../src/ReplaySession.cc:456
#7  0x00000000004ed9c9 in rr::ReplaySession::replay_step (this=0xea9ae0, constraints=...) at ../src/ReplaySession.cc:1477
#8  0x0000000000508789 in rr::ReplayTimeline::reverse_singlestep(rr::ReplayTimeline::Mark const&, rr::TaskishUid<rr::Task> const&, long, std::function<bool (rr::ReplayTask*)> const&, std::function<bool ()> const&) (
    this=<optimized out>, origin=..., step_tuid=..., step_ticks=<optimized out>, stop_filter=..., interrupt_check=...) at ../src/ReplayTimeline.cc:1190
#9  0x000000000050aeac in rr::ReplayTimeline::reverse_singlestep(rr::TaskishUid<rr::Task> const&, long, std::function<bool (rr::ReplayTask*)> const&, std::function<bool ()> const&) (this=0x7fff1d085eb8, tuid=..., tuid_ticks=31,
    stop_filter=..., interrupt_check=...) at ../src/ReplayTimeline.cc:1366
#10 0x000000000047ae43 in rr::GdbServer::debug_one_step (this=<optimized out>, last_resume_request=...) at ../src/GdbServer.cc:1205
#11 0x000000000047c66b in rr::GdbServer::serve_replay (this=<optimized out>, flags=...) at ../src/GdbServer.cc:1537
#12 0x00000000004e5af2 in rr::replay (trace_dir=..., flags=...) at ../src/ReplayCommand.cc:452
#13 rr::ReplayCommand::run (this=<optimized out>, args=...) at ../src/ReplayCommand.cc:567
#14 0x000000000054779e in main (argc=3, argv=0x7fff1d086798) at ../src/main.cc:237
```

## Reproduction steps

Download the [trace.zip](https://github.com/mozilla/rr/files/1502730/trace.zip) of the `simple_32` unit test and load it via `rr replay`. The following GDB commands will trigger the issue. 

```
hbreak * 0x70000003 
c
stepi
reverse-stepi
```

**Comment by rocallahan** [comment]
This testcase is trying to reverse-stepi over `rrcall_init_preload`, which is probably related to the root cause.

**Comment by rocallahan** [comment]
Part of the problem is that when we `stepi` from 0x70000003 instead of stopping at 0x70000005 after the syscall, we step through another instruction as well.

**Comment by rocallahan** [comment]
Another part of the problem is that instruction-execution hardware watchpoints are treated like breakpoints: the instruction is not executed. So in this case, where there is a hardware watchpoint at 0x70000003 and we singlestep starting there, we don't go anywhere. ReplayTimeline handles this for breakpoints but not for instruction-execution watchpoints.

**Comment by rocallahan** [comment]
Here's an interesting quirk on my Skylake laptop: I'm at 0x40070f with a hardware execution watchpoint at the next instruction, 0x400712. I do a singlestep. We stop at 0x400712 as expected, but *both* the BS bit and the B0 bit are set in DR6. B0 really shouldn't be set, because we shouldn't have tried to execute the instruction at 0x400712. No idea if this is a hardware issue or a kernel issue.

**Comment by gaasedelen** [comment]
> Part of the problem is that when we stepi from 0x70000003 instead of stopping at 0x70000005 after the syscall, we step through another instruction as well.

Yep, this is something I noticed around the same time I filed this issue. I should have created another issue for that, but I guess I forgot. It's interesting to see it might be related to this. 

Adding @khuey to the discussion as he proposed a fix back then for the fact that we skip an instruction after a syscall. While testing Kyle's quick attempt at a patch for the issue, I noticed a few regressions in unit tests and never got around to investigating it further.
